iotilebuild/iotile/build/config/site_scons/trub_script.py

```diff
@@ -27,51 +29,77 @@ def build_update_script(file_name, slot_assignments, os_info=None):
     env = Environment(tools=[])
+    files = []
 
-    slots = [_parse_slot(x[0]) for x in slot_assignments]
-    files = [ensure_image_is_hex(resolver.find_unique("firmware_image", x[1]).full_path) for x in slot_assignments]
+    if slot_assignments is not None:
+        slots = [_parse_slot(x[0]) for x in slot_assignments]
+        files = [ensure_image_is_hex(resolver.find_unique("firmware_image", x[1]).full_path) for x in slot_assignments]
+        env['SLOTS'] = slots
 
-    env['SLOTS'] = slots
     env['OS_INFO'] = os_info
+    env['APP_INFO'] = app_info
+
+    if sensor_graph is not None:
+        files.append(sensor_graph)
+
     env.Command([os.path.join('build', 'output', file_name)], files, action=Action(_build_reflash_script_action, "Building TRUB script at $TARGET"))
 
 
 def _build_reflash_script_action(target, source, env):
-    """Create a TRUB script containing tile and controller reflashes.
+    """Create a TRUB script containing tile and controller reflashes and/or sensorgraph
 
-    All of the files in source must be in intel hex format. This is guaranteed
+    If the app_info is provided, then the final source file will be a sensorgraph.
+    All subsequent files in source must be in intel hex format. This is guaranteed
     by the ensure_image_is_hex call in build_update_script.
     """
 
     out_path = str(target[0])
     source = [str(x) for x in source]
     records = []
-
-    for (controller, slot_id), image_path in zip(env['SLOTS'], source):
-        hex_data = IntelHex(image_path)
-        hex_data.padding = 0xFF
-
-        offset = hex_data.minaddr()
-        bin_data = bytearray(hex_data.tobinarray(offset, hex_data.maxaddr()))
-
-        if controller:
-            record = ReflashControllerRecord(bin_data, offset)
-        else:
-            record = ReflashTileRecord(slot_id, bin_data, offset)
-
-        records.append(record)
-
+    
+    #Update sensorgraph
+    app_info = env['APP_INFO']
+    if app_info is not None:
+        sensor_graph_file = source.pop(-1)
+        sensor_graph = compile_sgf(sensor_graph_file)
+        output = format_script(sensor_graph)
+        records += UpdateScript.FromBinary(output).records
+
+    #Update application firmwares
+    if env['SLOTS'] is not None:
+        for (controller, slot_id), image_path in zip(env['SLOTS'], source):
+            hex_data = IntelHex(image_path)
+            hex_data.padding = 0xFF
+
+            offset = hex_data.minaddr()
+            bin_data = bytearray(hex_data.tobinarray(offset, hex_data.maxaddr()))
+
+            if controller:
+                record = ReflashControllerRecord(bin_data, offset)
+            else:
+                record = ReflashTileRecord(slot_id, bin_data, offset)
+
+            records.append(record)
+
+    #Update App and OS Tag
     os_info = env['OS_INFO']
     if os_info is not None:
         os_tag, os_version = os_info
         records.append(SetDeviceTagRecord(os_tag=os_tag, os_version=os_version))
```

Comment:
You can do this with a single call to `SetDeviceTagRecord` with all 4 parameters set or None if you don't want to specify the app or os part of the tag.

https://github.com/iotile/coretools/blob/85e7caa3a31f1b59c6fbe5b7ee7f47d731399d47/iotilecore/iotile/core/hw/update/records/set_version.py#L22